keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1346 @@
+- Policy definitions (MutatingAdmissionPolicy)
+- Policy bindings (MutatingAdmissionPolicyBinding)
+- Policy param resources (custom resources or config maps)
+
+The idea is to leverage the CEL power of the object construction and allow users to define how they want to mutate the admission request through CEL expression.
+This proposal aims to allow mutations to be expressed using the "apply configuration" introduced by Server Side Apply. 
+And users would be able to define only the fields they care about inside MutatingAdmissionPolicy, the object will be constructed using CEL which would be similar to a Server Side Apply configuration patch and then be merged into the request object using the structural merge strategy. 
+See sigs.k8s.io/structured-merge-diff for more details.
+
+Note: See the alternative consideration section for the alternatives.
+
+Pros:
+- Consistent with Server Side Apply so that we will continue investing SSA as the best way to do patch updates to resources;
+- Does not require the users to learn a new syntax;
+- Inherit the declarative nature;
+- Existing merging strategy to be leverage without rewritten and the effort of maintaining multiple systems;
+- Support the existing makers/openapi extensions.
+
+Cons:
+- Lack of deletion support (see the unsetting values section for the details and workaround);
+- All operations of SSA require the schemas of the incoming objects, which require the schemas to be available from the discovery endpoint or compiled-in. 
+  - For API extensions types like CRDs, this can cause a delay before the policy can start enforcing requests.
+  
+
+### Phase 1
+
+#### API Shape
+Similar to the `validations` field in ValidatingAdmissionPolicy, a `mutations` field will be defined inside MutatingAdmissionPolicy which allows users to define a list of mutations that apply to the specific resources.
+Each mutation field contains a CEL expression which evaluates to a partially populated object representing a Server Side Apply "apply configuration". The apply configuration is then merged into the request object.
+
+Here is an example of injecting an initContainer.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  paramKind:
+    group: mutations.example.com
+    kind: Sidecar
+    version: v1
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == params.name)"
+  failurePolicy: Fail
+  reinvocationPolicy: IfNeeded
+  mutations:
+    - mutation: >
+        Object{
+          spec: Object.spec{
+            initContainers: [
+              Object.spec.initContainers{
+                name: params.name,
```

Comment:
if the intent is to add a new container, but we're using the SSA merge logic, then if this name collides with a name already present in the initContainers, then this would overwrite it.  That would be unexpected, right?

Reply:
That is why we have the `matchConditions` above to avoid those conflict. So user will check if the same name exists before mutation.

Reply:
> That is why we have the `matchConditions` above to avoid those conflict. So user will check if the same name exists before mutation.

How would this author achieve a goal of, "every pod has an init container that does thing/X"?  Just hoping conflicting container names do the same thing?

Reply:
If there is a existing initContainer with the same name, yes it will be overwritten. This is the same behavior in existing mutation webhook. The workaround would be having two mutation policy setup with matchConditions check "if there is an existing init container with the same name or not". That will result in having two mutation policy setup which is not ideal. Or we could have a matchCondition alike field attached with each mutation. What would you think?